Plutarch/Lift.hs

```diff
@@ -1,109 +1,70 @@
 
-  -- {-
-  -- Convert a Plutarch term to the associated Haskell value. Fail otherwise.
-  -- This will fully evaluate the arbitrary closed expression, and convert the
-  -- resulting value.
-  -- -}
-  plift' :: ClosedTerm p -> Either LiftError (PHaskellType p)
+type PLift = PUnsafeLiftDecl
 
--- | Like `pconstant'` but TypeApplication-friendly
-pconstant :: forall p s. PLift p => PHaskellType p -> Term s p
-pconstant = pconstant'
+pconstant :: forall p s. PLift p => PLifted p -> Term s p
+pconstant x = punsafeConstantInternal $ PLC.someValue @(PConstantRepr (PLifted p)) @PLC.DefaultUni $ pconstantToRepr x
 
--- | Like `plift'` but fails on error.
-plift :: (PLift p, HasCallStack) => ClosedTerm p -> PHaskellType p
-plift prog = either (error . show) id $ plift' prog
-
-{- | DerivingVia representation to auto-derive `PLift` for Plutarch types
- representing builtin Plutus types in the `DefaultUni`.
-
- The `h` parameter is the Haskell type associated with `p`. Example use:
-
- > deriving PLift via (PBuiltinType PInteger Integer)
-
- See instance below.
--}
-newtype PBuiltinType (p :: k -> Type) (h :: Type) s = PBuiltinType (p s)
-
-instance
-  ( PLC.KnownTypeIn PLC.DefaultUni (UPLC.Term PLC.DeBruijn PLC.DefaultUni PLC.DefaultFun ()) h
-  , PLC.DefaultUni `PLC.Contains` h
-  ) =>
-  PLift (PBuiltinType p h)
-  where
-  type PHaskellType (PBuiltinType p h) = h
-  pconstant' =
-    punsafeConstantInternal . PLC.Some . PLC.ValueOf (PLC.knownUniOf (Proxy @h))
-  plift' prog =
-    case evaluateScript (compile prog) of
-      Left e -> Left $ LiftError_ScriptError e
-      Right (_, _, Scripts.unScript -> UPLC.Program _ _ term) ->
-        first (LiftError_EvalException . showEvalException) $
-          readKnownSelf term
+-- | Error during script evaluation.
+data LiftError
+  = LiftError_ScriptError Scripts.ScriptError
+  | LiftError_EvalException (ErrorWithCause (MachineError PLC.DefaultFun) ())
+  | LiftError_FromRepr
+  | LiftError_WrongRepr
+  deriving stock (Eq, Show)
 
-showEvalException :: EvaluationException CekUserError (MachineError PLC.DefaultFun) (UPLC.Term UPLC.DeBruijn PLC.DefaultUni PLC.DefaultFun ()) -> Text
-showEvalException = T.pack . show
+plift' :: forall p. PUnsafeLiftDecl p => ClosedTerm p -> Either LiftError (PLifted p)
+plift' prog = case evaluateScript (compile prog) of
+  Right (_, _, Scripts.unScript -> UPLC.Program _ _ term) ->
+    case readKnownConstant @_ @(PConstantRepr (PLifted p)) @(MachineError PLC.DefaultFun) Nothing term of
+      Right r -> case pconstantFromRepr r of
+        Just h -> Right h
+        Nothing -> Left LiftError_FromRepr
+      Left e -> Left $ LiftError_EvalException e
+  Left e -> Left $ LiftError_ScriptError e
+
+plift :: forall p. (HasCallStack, PLift p) => ClosedTerm p -> (PLifted p)
+plift prog = case plift' prog of
+  Right x -> x
+  Left e -> error $ "plift failed: " <> show e
+
+newtype DerivePConstantViaCoercible (h :: Type) (p :: PType) (r :: Type) = DerivePConstantViaCoercible h
```

Comment:
Is the `h` used for `DerivePConstantViaCoercible`? Is there an example usage somewhere perhaps?

Reply:
The `h` is used because it's derived *for* the `h`. `PInteger` uses it I think.

Reply:
Indeed, albeit I still think it's useful to have the representation be separate.

Reply:
I don't think this is important enough to change.